server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexUpgradeService.java

```diff
@@ -69,11 +76,15 @@ public MetadataIndexUpgradeService(NodeContext nodeContext,
      * If the index does not need upgrade it returns the index metadata unchanged, otherwise it returns a modified index metadata. If index
      * cannot be updated the method throws an exception.
      */
-    public IndexMetadata upgradeIndexMetadata(IndexMetadata indexMetadata, IndexTemplateMetadata indexTemplateMetadata, Version minimumIndexCompatibilityVersion) {
+    public IndexMetadata upgradeIndexMetadata(IndexMetadata indexMetadata,
+                                              IndexTemplateMetadata indexTemplateMetadata,
+                                              Version minimumIndexCompatibilityVersion,
+                                              UserDefinedFunctionsMetadata userDefinedFunctionsMetadata) {
         // Throws an exception if there are too-old segments:
         if (isUpgraded(indexMetadata)) {
             return indexMetadata;
         }
+        userDefinedFunctionService.updateImplementations(userDefinedFunctionsMetadata.functionsMetadata());
```

Comment:
Maybe add a check if the UDFs were ever set?

Reply:
I don't get this question, could you please provide more details?